lib/kube/proxy/transport_retryable.go

```diff
@@ -0,0 +1,274 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package proxy
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"log/slog"
+	"net/http"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// retryableTransport wraps an http.RoundTripper to enable automatic retry
+// on HTTP/2 GOAWAY frames by buffering request bodies in memory.
+//
+// Kubernetes API servers send GOAWAY on HTTP/2 connections to redistribute
+// load across replicas. Go's http2.Transport automatically retries if
+// req.GetBody is set.
+//
+// A global semaphore limits total buffered memory to 500 MiB across all
+// concurrent requests, preventing OOM. Individual requests have a maximum
+// body size of 50 MiB. Requests are blocked if required memory cannot be
+// acquired from the semaphore. These limits are tunable with environment
+// variables.
+//
+// Requests that are not retried are:
+// 1. HTTP/1.1 protocol upgrades (exec/attach/portforward), which won't
+// receive an HTTP/2.0 GOAWAY.
+// 2. Requests with unknown body sizes, used with chunked encoding,
+// cannot safely acquire the semaphore.
+type retryableTransport struct {
+	inner               http.RoundTripper
+	log                 *slog.Logger
+	semaphore           *semaphore.Weighted
+	maxBufferPerRequest int64
+}
+
+// newRetryableTransport enables automatic retry during HTTP/2 GOAWAY
+// load balancing.
+//
+// Intended to wrap transports created via newH2Transport() or equivalent
+// HTTP/2-configured transports. Wrapping non-HTTP/2 transports would add
+// unnecessary buffering.
+func newRetryableTransport(inner http.RoundTripper, log *slog.Logger, semaphore *semaphore.Weighted, maxBufferPerRequest int64) *retryableTransport {
+	return &retryableTransport{
+		inner:               inner,
+		log:                 log,
+		semaphore:           semaphore,
+		maxBufferPerRequest: maxBufferPerRequest,
+	}
+}
+
+// RoundTrip implements http.RoundTripper and makes requests retryable.
+func (rt *retryableTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Detect HTTP/1.1 protocol upgrades that won't receive HTTP/2 GOAWAY.
+	if rt.isUpgrade(req) {
+		// Perform roundtrip, no retry available.
+		return rt.inner.RoundTrip(req)
+	}
+
+	if req.Body == nil || req.GetBody != nil {
+		// Perform roundtrip, retry available.
+		return rt.inner.RoundTrip(req)
+	}
+
+	// Buffer body to make retryable.
+	releaseSemaphore := rt.makeRetryable(req)
+
+	// Release semaphore after roundtrip completes.
+	defer releaseSemaphore()
+
+	// Perform roundtrip, retry available.
+	return rt.inner.RoundTrip(req)
+}
+
+// makeRetryable buffers the request body to enable retry on HTTP/2 GOAWAY.
+func (rt *retryableTransport) makeRetryable(req *http.Request) func() {
+	// Chunked request are not retried due to unknown size.
+	if req.ContentLength < 0 {
```

Comment:
What's the distinction between the checks done here and the checks done in RoundTrip?